internal/controllers/utils/events.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"sort"
+	"time"
+)
+
+const (
+	// Generic Event states
+	// InProgress indicates a in progress state
+	InProgress = "InProgress"
+	// InProgress indicates a completed state
+	Completed = "Completed"
+)
+
+// Event is the generic representation of an event
+type Event struct {
+	ObjectID  string
+	Timestamp time.Time
+	State     string
+}
+
+// EventHistory contains a slice of events
+type EventHistory struct {
+	History []*Event
+}
+
+// IsTimedOut proceses a process history an determines the overall state implied
+// by the events.
+// returns true if for any ObjectID:
+// - the overall state was InProgress from the initial time for more than timeout
+// - the overall state was InProgress from the last transition to InProgress for more than timeout
```

Comment:
Let's continue the discussion in https://github.com/openshift-kni/oran-o2ims/pull/246#discussion_r1803846429. In my understanding it is not the same since the timeout will be different based on the history.